include/cpr/singleton.h

```diff
@@ -10,38 +11,33 @@
 #endif
 
 #ifndef CPR_SINGLETON_DECL
-#define CPR_SINGLETON_DECL(Class) \
-  public:                         \
-    static Class* GetInstance();  \
-    static void ExitInstance();   \
-                                  \
-  private:                        \
-    CPR_DISABLE_COPY(Class)       \
-    static Class* s_pInstance;    \
-    static std::mutex s_mutex;
+#define CPR_SINGLETON_DECL(Class)    \
+  public:                            \
+    static Class* GetInstance();     \
+    static void ExitInstance();      \
+                                     \
+  private:                           \
+    CPR_DISABLE_COPY(Class)          \
+    static Class* s_pInstance;       \
+    static std::once_flag s_getFlag; \
+    static std::once_flag s_exitFlag;
 #endif
 
 #ifndef CPR_SINGLETON_IMPL
-#define CPR_SINGLETON_IMPL(Class)         \
-    Class* Class::s_pInstance = nullptr;  \
-    std::mutex Class::s_mutex;            \
-    Class* Class::GetInstance() {         \
-        if (s_pInstance == nullptr) {     \
-            s_mutex.lock();               \
-            if (s_pInstance == nullptr) { \
-                s_pInstance = new Class;  \
-            }                             \
-            s_mutex.unlock();             \
-        }                                 \
-        return s_pInstance;               \
-    }                                     \
-    void Class::ExitInstance() {          \
-        s_mutex.lock();                   \
-        if (s_pInstance) {                \
-            delete s_pInstance;           \
-            s_pInstance = nullptr;        \
-        }                                 \
-        s_mutex.unlock();                 \
+#define CPR_SINGLETON_IMPL(Class)                                            \
+    Class* Class::s_pInstance = nullptr;                                     \
+    std::once_flag Class::s_getFlag{};                                       \
+    std::once_flag Class::s_exitFlag{};                                      \
+    Class* Class::GetInstance() {                                            \
+        std::call_once(Class::s_getFlag, []() { s_pInstance = new Class; }); \
+        return s_pInstance;                                                  \
+    }                                                                        \
+    void Class::ExitInstance() {                                             \
+        std::call_once(Class::s_exitFlag, []() {                             \
+            assert(s_pInstance != 0);                                        \
```

Comment:
```suggestion
            assert(s_pInstance0);                                            \
```

More cpr like null pointer check.